File: sorting/java/MergeSort.java

class MergeSort implements SortAlgorithm
{
	public void sort(int[] array)
	{
		int[] tmp = new int[array.length];
		m_sort(array, tmp, 0, array.length);
	}
	private void m_sort(int[] array, int[] tmp, int start, int end)
	{
		if(end - start > 1)
		{
			int middle = (start + end) / 2;
			m_sort(array, tmp, start, middle);
			m_sort(array, tmp, middle, end);
			merge(array, tmp, start, middle, end);
		}
	}
	private void merge(int[] array, int[] tmp, int start, int middle, int end)
	{
		int i = start, j = middle, k = 0;
		while(i < middle && j < end)
		{
			if(array[i] < array[j])
				tmp[ k++ ] = array[ i++ ];
			else
				tmp[ k++ ] = array[ j++ ];
		}
		while(i < middle)
			tmp[ k++ ] = array[ i++ ];
		while(j < end)
			tmp[ k++ ] = array[ j++ ];
		for(i = 0; i < k; i++)
			array[start + i] = tmp[i];
	}
}

File: sorting/perl/merge_sort.pl
package sorting;
use Exporter;
@ISA = ('Exporter');
@EXPORT = ('merge_sort');

sub merge_sort {
	my $array = shift @_;
	m_sort($array, 0, $$array);
}

sub m_sort {
	my ($array, $start, $end) = @_;
	if($end - $start > 1) {
		my $middle = int(($start + $end) / 2);
		m_sort($array, $start, $middle);
		m_sort($array, $middle, $end);
		merge($array, $start, $middle, $end);
	}
}

sub merge {
	my ($array, $start, $middle, $end) = @_;
	my @tmp;
	my ($i, $j) = ($start, $middle);
	while($i < $middle and $j < $end) {
		if($$array[$i] < $$array[$j]) {
			push @tmp, $$array[$i++];
		} else {
			push @tmp, $$array[$j++];
		}
	}
	while($i < $middle) {
		push @tmp, $$array[$i++];
	}
	while($j < $end) {
		push @tmp, $$array[$j++];
	}
	for(my $i = $start; $i < $end; $i++) {
		$$array[$i] = $tmp[$i - $start];
	}
}

1;
